### PR TITLE
Refactor ModuleMaker into ModuleMaker and LibraryMaker

### PR DESCRIPTION
**Pull request by muellerzr**
# Introduce a `LibraryMaker` class

## What does this add?

This PR refactors the `ModuleMaker` class into a much more barebones version of itself, and renames what was ModuleMaker into a `LibraryMaker` class

## Why is it needed?

As I was trying to write my pytest integration runner, I was hitting `ast.parse` errors and I couldn't quite figure out why. It turned out to be because of checking for and writing the `__all__` in a notebook + the future imports, but there are times when I don't care about this happening at all (such as exporting tests to a test file). 

As a result, I figured out that we can separate `ModuleMaker` into even simpler components, allowing for even more flexible use of this framework.

### User-facing:

A new `LibraryMaker` class was introduced and `ModuleMaker` was simplified. Any custom implementations that want to emulate or tweak the behavior of the current `ModuleMaker` class can be implemented through sub-classing the `ModuleMaker` class. 

The base class will write the `AUTOGENERATED` comment and any of the cells you wish to write. If calling the base `.write` implementation multiple times, a newline flag can be turned `False` so a `\n` isn't added to the end of the file call. 

## When would I use it, and when wouldn't I?

This is for users who want to create their own Makers that don't follow the base idea of nbdev, (e.g. focus on **just** a single notebook to a single .py source code) and instead want the flexibility to create a variety of different file types/contents based on the contents in each flagged cell, with the possibility of edits occurring so `ast.parse` may not be desired. 

## Does a similar feature exist? If so, why is this better?

I originally tried implementing this as its own `ModuleMaker` subclass, but there was a large chunk of duplicated code that could be simplified by subclassing and refactoring.

### CODE REVIEW

File: nbs/02_maker.ipynb [hunk 1]
```diff
@@ -180,15 +180,34 @@
    "source": [
```
Comment (jph00) [inline nbs/02_maker.ipynb:180]: <p>These are being exported, but are not documented. Either add docstrings and markdown prose to describe them, or make them private. If you make them private, remove or hide the test below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbprocess/pull/73/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (jph00) [inline nbs/02_maker.ipynb:180]: <p>I don't think this adds anything to the docstring, so remove it (and update the docstring if you feel it's unclear)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbprocess/pull/73/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

**Comment by jph00** [comment]
@muellerzr I like the direction here -- thanks for the PR! It looks like this refactoring has ended up with more code, but normally I'd expect refactoring to reduce the amount of code. Is there any way you could simplify, or move stuff that's only needed for your project over to there?

PS: don't forget to at-mention me on PRs so I don't miss them!

**Comment by muellerzr** [comment]
@jph00 moved it around as much as I could, now there's not *too* many additions but I'm also happy long-term with what this accomplishes: letting us only call `maker.make()` once.

File: nbs/02_maker.ipynb (hunk 1 above)
Comment (muellerzr) [reply nbs/02_maker.ipynb:180]: <p>Fixed.</p><p> </p>

Comment (muellerzr) [reply nbs/02_maker.ipynb:180]: <p>Fixed. </p>

**Comment by muellerzr** [comment]
@jph00 closing this then opening another one. There's an easier refactor, you are right, and tested this on a new branch of mine.

**Pull request closed without merging**